jvmhs/src/Jvmhs/Analysis/Reduce.hs

```diff
@@ -92,3 +93,100 @@ findUsedClasses  =
           (traverse.classNames)
           (traverse.classNames)
           nothing) cls
+
+
+chopVector :: Int -> V.Vector x -> V.Vector (V.Vector x)
+chopVector n vec
+  | divisible = chopVector' blockSize n vec V.empty
+  | otherwise = chopVector' blockSize (vectorLength `mod` n) vec V.empty
+  where blockSize = vectorLength `roundUpQuot` n
+        vectorLength = length vec
+        divisible =  length vec `mod` n == 0
+
+chopVector' ::
+     Int
+  -- ^ size of each Vec (delta)
+  -> Int
+  -- ^ num of vec remaining should be larger
+  -> V.Vector x
+  -> V.Vector (V.Vector x)
+  -> V.Vector (V.Vector x)
+
+chopVector' i n vec rslt
+  | V.null vec        = rslt
+  | otherwise         = let (block, rest) = V.splitAt newBlockSize vec
+                          in chopVector'
+                            newBlockSize newN rest (snoc rslt block)
+  where newN         = if n > -1 then n - 1 else n
+        newBlockSize = if n == 0 then i-1 else i
+
+roundUpQuot :: Int -> Int -> Int
+roundUpQuot i j = ceiling ((fromIntegral i / fromIntegral j)::Float)
+
+deltaI ::
+     V.Vector (V.Vector x)
+  -> Int
+  -> V.Vector x
+deltaI vecs i =
+  vecs V.! i
+
+deltaComplementI ::
+    V.Vector (V.Vector x)
+  -> Int
+  -> V.Vector x
+deltaComplementI vecs i =
+  V.foldl1 (V.++) $ V.ifilter (\j _ -> i /= j) vecs
+
+
+ddmin ::
+     Monad m
+--     (Monad m, MonadIO m, Show x)
+  => V.Vector x
+  -> (V.Vector x -> m Bool)
+  -- ^ a function can test delta
+  -> m(V.Vector x)
+
+ddmin world = ddmin' world 2
+
+ddmin' ::
+     Monad m
+--     (Monad m, MonadIO m, Show x)
+  => V.Vector x
+  -> Int
+  -> (V.Vector x -> m Bool)
+  -- ^ a function can test delta
+  -> m(V.Vector x)
+ddmin' world n testFunc = do
+  deltaTrueVec <- V.filterM testFunc deltaVec
+  deltaComplTrueVec <- V.filterM testFunc deltaComplVec
+  if not $ null deltaTrueVec
+    then
+      ddmin' (V.head deltaTrueVec) 2 testFunc
+    else if not $ null deltaComplTrueVec
+      then
+        ddmin' (V.head deltaComplTrueVec) (max (n-1) 2) testFunc
+      else if n < V.length world
+        then
+           ddmin' world (min (V.length world) (2*n)) testFunc
+        else
+           return world
+
+  where enumVec       = V.fromList [0..(n-1)]
+        vecs          = chopVector n world
+        deltaVec      = V.map (deltaI vecs) enumVec
```

Comment:
This is just iterating through the list? deltaVec = vecs?

Reply:
Right. I should change it to vecs.